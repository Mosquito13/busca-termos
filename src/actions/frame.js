import actionTypes from '../constants/actionTypes';
import apiUtils from '../utils/apiUtils';

export const setMaximized = (value) => ({
  type: actionTypes.SET_MAXIMIZED,
  value
});

export const close = () => () => apiUtils.close();

export const minimize = () => () => apiUtils.minimize();

export const maximize = () => () => apiUtils.maximize();

export const unmaximize = () => () => apiUtils.unmaximize();

export const registerFrameListeners = () => (dispatch) => {
  apiUtils.registerMaximizeListener(() => {
    dispatch(setMaximized(true));
  });

  apiUtils.registerUnmaximizeListener(() => {
    dispatch(setMaximized(false));
  });
};
