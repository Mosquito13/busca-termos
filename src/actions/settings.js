import apiUtils from '../utils/apiUtils';
import actionTypes from '../constants/actionTypes';
import storageUtils from '../utils/storageUtils';
import coreActions from './core';

const settingsActions = {
  setValid(value) {
    return {
      type: actionTypes.SET_VALID,
      value
    };
  },

  setValidating(value) {
    return {
      type: actionTypes.SET_VALIDATING,
      value
    };
  },

  setLanguageFolder(value) {
    return {
      type: actionTypes.SET_LANGUAGE_FOLDER,
      value
    };
  },

  setLanguageFolderError(value) {
    return {
      type: actionTypes.SET_LANGUAGE_FOLDER_ERROR,
      value
    };
  },

  setMainLanguage(value) {
    return {
      type: actionTypes.SET_MAIN_LANGUAGE,
      value
    };
  },

  toggleTranslation(id, value) {
    return {
      type: actionTypes.TOGGLE_TRANSLATION,
      id,
      value
    };
  },

  toggleCompactLayout(value) {
    return {
      type: actionTypes.TOGGLE_COMPACT_LAYOUT,
      value
    };
  },

  toggleDarkTheme(value) {
    return {
      type: actionTypes.TOGGLE_DARK_THEME,
      value
    };
  },

  setSettings(value) {
    return {
      type: actionTypes.SET_SETTINGS,
      value
    };
  },

  validateAndSaveFirstSettings(languageFolder) {
    return async (dispatch) => {
      dispatch(this.setValidating(true));

      const isValid = await apiUtils.validateLanguageFolder(languageFolder);

      if (!isValid) {
        dispatch(this.setLanguageFolderError('Caminho invÃ¡lido.'));
        dispatch(this.setValid(false));
      } else {
        dispatch(coreActions.setLoading(true));
        dispatch(this.setValid(true));
        dispatch(this.saveSettings({ languageFolder }));
      }

      dispatch(this.setValidating(false));
    };
  },

  loadSettings() {
    return async (dispatch) => {
      const settings = storageUtils.getSettings();
      const isValid = await apiUtils.validateLanguageFolder(settings?.languageFolder);

      if (!isValid) {
        dispatch(this.setValid(false));
        dispatch(coreActions.setLoading(false));
      } else {
        dispatch(this.setSettings(settings));
        dispatch(coreActions.loadData(settings));
      }
    };
  },

  saveSettings(values) {
    return {
      type: actionTypes.SAVE,
      values
    };
  }
};

export default settingsActions;
