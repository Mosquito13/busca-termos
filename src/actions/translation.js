import actionTypes from '../constants/actionTypes';
import coreSelectors from '../selectors/core';

export const setTranslation = (id, value) => ({
  type: actionTypes.SET_TRANSLATION,
  id,
  value
});

export const loadTranslation = (id, termId) => (dispatch, getState) => {
  const state = getState();
  const data = coreSelectors.getData(state);
  const translation = data[id].find((d) => d.id === termId);

  dispatch(setTranslation(id, translation?.content));
};
