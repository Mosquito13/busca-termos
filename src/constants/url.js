const url = {
  LIST_APP_RELEASES: 'https://api.github.com/repos/Mosquito13/busca-termos/releases',
  ISSUES_PAGE: 'https://github.com/Mosquito13/busca-termos/issues'
};

export default url;
