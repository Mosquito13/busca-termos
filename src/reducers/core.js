import actionTypes from '../constants/actionTypes';

const initial = {
  loading: true,
  data: null,
  filter: '',
  selectedItemId: null,
  version: '',
  appHasUpdate: null,
  updateURL: ''
};

const coreReducer = (state = initial, action) => {
  switch (action.type) {
    case actionTypes.SET_LOADING:
      return { ...state, loading: action.loading };

    case actionTypes.SET_DATA:
      return { ...state, data: action.data };

    case actionTypes.SET_FILTER:
      return { ...state, filter: action.filter };

    case actionTypes.SET_APP_VERSION:
      return { ...state, version: action.version };

    case actionTypes.SET_APP_HAS_UPDATE:
      return { ...state, appHasUpdate: action.appHasUpdate, updateURL: action.updateURL };

    case actionTypes.SET_SELECTED_ITEM_ID:
      return { ...state, selectedItemId: action.id };

    default:
      return state;
  }
};

export default coreReducer;
