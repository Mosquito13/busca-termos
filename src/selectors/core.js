const getSlice = (state) => state.Core;

const coreSelectors = {
  isLoading(state) {
    return getSlice(state).loading;
  },

  getData(state) {
    return getSlice(state).data;
  },

  getFilter(state) {
    return getSlice(state).filter;
  },

  getAppVersion(state) {
    return getSlice(state).version;
  },

  getAppHasUpdate(state) {
    return getSlice(state).appHasUpdate;
  },

  getUpdateURL(state) {
    return getSlice(state).updateURL;
  },

  getSelectedItemId(state) {
    return getSlice(state).selectedItemId;
  }
};

export default coreSelectors;
