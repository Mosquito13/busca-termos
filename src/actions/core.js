import axios from 'axios';

import actionTypes from '../constants/actionTypes';
import releaseUtils from '../utils/releaseUtils';
import coreSelectors from '../selectors/core';
import apiUtils from '../utils/apiUtils';
import url from '../constants/url';

export const setLoading = (loading) => ({
  type: actionTypes.SET_LOADING,
  loading
});

export const setData = (data) => ({
  type: actionTypes.SET_DATA,
  data
});

export const setSelectedItemId = (id) => ({
  type: actionTypes.SET_SELECTED_ITEM_ID,
  id
});

export const setFilter = (filter) => ({
  type: actionTypes.SET_FILTER,
  filter
});

export const setAppVersion = (version) => ({
  type: actionTypes.SET_APP_VERSION,
  version
});

export const setAppHasUpdate = (appHasUpdate, updateURL) => ({
  type: actionTypes.SET_APP_HAS_UPDATE,
  appHasUpdate,
  updateURL
});

export const checkForUpdates =
  (currentVersion) => async (dispatch, getState) => {
    const state = getState();
    const hasUpdate = coreSelectors.getAppHasUpdate(state);

    if (hasUpdate === null) {
      const releases = await axios.get(url.LIST_APP_RELEASES);
      const [latestReleaseVersion, latestReleaseURL] =
        releaseUtils.parseLastRelease(releases);

      const hasUpdate = releaseUtils.hasUpdate(
        currentVersion,
        latestReleaseVersion
      );

      dispatch(setAppHasUpdate(hasUpdate, latestReleaseURL));
    }
  };

export const loadData = (settings) => async (dispatch) => {
  const data = await apiUtils.loadData(settings?.languageFolder);
  const appVersion = await apiUtils.loadAppVersion();

  dispatch(setAppVersion(appVersion));
  dispatch(checkForUpdates(appVersion));
  dispatch(setData(data));
  dispatch(setLoading(false));
};

export const openBrowserWithURL = (url) => async () => {
  apiUtils.openBrowserWithURL(url);
};
