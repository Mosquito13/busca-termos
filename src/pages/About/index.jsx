import { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { FiExternalLink, FiCheck, FiRefreshCcw } from 'react-icons/fi';
import classNames from 'classnames';

import Link from '../../components/common/Link';
import Button from '../../components/common/Button';
import Heading1 from '../../components/common/Heading1';
import Paragraph from '../../components/common/Paragraph';
import Icon, { RESPONSIVE } from '../../components/common/Icon';

import url from '../../constants/url';
import coreActions from '../../actions/core';
import coreSelectors from '../../selectors/core';

import './styles.scss';

const About = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const appVersion = useSelector(coreSelectors.getAppVersion);
  const appHasUpdate = useSelector(coreSelectors.getAppHasUpdate);
  const lastUpdateLink = useSelector(coreSelectors.getUpdateURL);

  const onClickIssueLink = useCallback(() => {
    dispatch(coreActions.openBrowserWithURL(url.ISSUES_PAGE));
  }, [dispatch]);

  const onClickUpdateLink = useCallback(() => {
    dispatch(coreActions.openBrowserWithURL(lastUpdateLink));
  }, [dispatch, lastUpdateLink]);

  const onClickClose = useCallback(() => {
    navigate(-1);
    dispatch(coreActions.setLoading(true));
  }, [navigate, dispatch]);

  const classesIconUpdate = classNames(
    'about__info-updates-icon',
    appHasUpdate && 'about__info-updates-icon--red'
  );

  return (
    <div className="about">
      <div className="about__title">
        <Heading1 value="BuscaTermos" />
        <div className="about__title-date">
          <Paragraph value={`2014 - ${new Date().getFullYear()}`} />
        </div>
      </div>
      <div className="about__info">
        <div className="about__info-item">
          <Paragraph value="Versão:" bold />
          <Paragraph value={appVersion} />
        </div>
        <div className="about__info-item">
          <Paragraph value="Atualizações:" bold />
          <div className="about__info-updates">
            <div className={classesIconUpdate}>
              <Icon size={RESPONSIVE}>
                {appHasUpdate ? <FiRefreshCcw /> : <FiCheck />}
              </Icon>
            </div>
            {appHasUpdate ? (
              <Link
                value="Existem atualizações disponíveis"
                icon={<FiExternalLink />}
                onClick={onClickUpdateLink}
              />
            ) : (
              <Paragraph value="O aplicativo está atualizado" />
            )}
          </div>
        </div>
        <div className="about__info-item">
          <Paragraph value="Desenvolvido por:" bold />
          <Paragraph value="Giovane de Oliveira" />
          <Paragraph value="giovane.oliveira@softexpert.com" />
        </div>
        <div className="about__info-item">
          <Paragraph value="Bugs e/ou melhorias? Cadastre uma issue em:" bold />
          <Link
            value="BuscaTermos - Issues"
            icon={<FiExternalLink />}
            onClick={onClickIssueLink}
          />
        </div>
      </div>
      <div className="about__button">
        <Button text="Voltar" onClick={onClickClose} />
      </div>
    </div>
  );
};

export default About;
