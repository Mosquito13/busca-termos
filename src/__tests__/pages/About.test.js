import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import * as reactRedux from 'react-redux';

import About from '../../pages/About';
import url from '../../constants/url';
import * as coreActions from '../../actions/core';

const navigateMock = jest.fn();
const reactRouterDomMock = { useNavigate: () => navigateMock };

jest.mock('react-router-dom', () => reactRouterDomMock);

const getStateMock = (appHasUpdate) => ({
  Core: {
    version: '1.6.9',
    appHasUpdate,
    updateURL: 'anURL/latest'
  },
  Settings: {
    compactLayout: false
  }
});

const prepareMocks = (dispatchMock, stateMock) => {
  jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock);
  jest
    .spyOn(reactRedux, 'useSelector')
    .mockImplementation((selector) => selector(stateMock));
};

describe(`${About.name}`, () => {
  describe('tests', () => {
    let wrapper, dispatchMock;

    beforeAll(() => {
      dispatchMock = jest.fn();

      jest.spyOn(coreActions, 'openBrowserWithURL');
      jest.spyOn(coreActions, 'setLoading');
    });

    it('should render without app update', () => {
      prepareMocks(dispatchMock, getStateMock(false));

      wrapper = shallow(<About />);

      expect(wrapper.find('.about')).toHaveLength(1);

      expect(wrapper.find('.about__info-updates-icon')).toHaveLength(1);
      expect(wrapper.find('.about__info-updates-icon--red')).toHaveLength(0);

      expect(wrapper.find('[data-testid="link-update"]')).toHaveLength(0);
      expect(wrapper.find('[data-testid="p-app-updated"]')).toHaveLength(1);
    });

    it('should open issues page', () => {
      wrapper.find('[data-testid="link-issues"]').simulate('click');

      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(
        coreActions.openBrowserWithURL(url.ISSUES_PAGE)
      );
    });

    it('should render with app update', () => {
      wrapper.unmount();

      prepareMocks(dispatchMock, getStateMock(true));

      wrapper = shallow(<About />);

      expect(wrapper.find('.about')).toHaveLength(1);

      expect(wrapper.find('.about__info-updates-icon')).toHaveLength(1);
      expect(wrapper.find('.about__info-updates-icon--red')).toHaveLength(1);

      expect(wrapper.find('[data-testid="link-update"]')).toHaveLength(1);
      expect(wrapper.find('[data-testid="p-app-updated"]')).toHaveLength(0);
    });

    it('should open update page', () => {
      wrapper.find('[data-testid="link-update"]').simulate('click');

      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(
        coreActions.openBrowserWithURL('anURL/latest')
      );
    });

    it('should navigate back when close', () => {
      wrapper.find('[data-testid="btn-close"]').simulate('click');

      expect(navigateMock).toBeCalledTimes(1);
      expect(navigateMock).toBeCalledWith(-1);

      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(coreActions.setLoading(true));
    });
  });

  describe('snapshots', () => {
    it('should render without app update', () => {
      prepareMocks(jest.fn(), getStateMock(false));

      const tree = renderer.create(<About />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with app update', () => {
      prepareMocks(jest.fn(), getStateMock(true));

      const tree = renderer.create(<About />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
