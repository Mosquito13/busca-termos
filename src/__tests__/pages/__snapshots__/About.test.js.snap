// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About snapshots should render with app update 1`] = `
<div
  className="about"
>
  <div
    className="about__title"
  >
    <h1
      className="heading-1"
    >
      BuscaTermos
    </h1>
    <div
      className="about__title-date"
    >
      <p
        className="paragraph"
      >
        2014 - 2022
      </p>
    </div>
  </div>
  <div
    className="about__info"
  >
    <div
      className="about__info-item"
    >
      <p
        className="paragraph paragraph--bold"
      >
        Versão:
      </p>
      <p
        className="paragraph"
      >
        1.6.9
      </p>
    </div>
    <div
      className="about__info-item"
    >
      <p
        className="paragraph paragraph--bold"
      >
        Atualizações:
      </p>
      <div
        className="about__info-updates"
      >
        <div
          className="about__info-updates-icon about__info-updates-icon--red"
        >
          <div
            className="icon"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="1 4 1 10 7 10"
              />
              <polyline
                points="23 20 23 14 17 14"
              />
              <path
                d="M20.49 9A9 9 0 0 0 5.64 5.64L1 10m22 4l-4.64 4.36A9 9 0 0 1 3.51 15"
              />
            </svg>
          </div>
        </div>
        <button
          className="link"
          onClick={[Function]}
        >
          <div
            className="link__text"
          >
            Existem atualizações disponíveis
          </div>
          <div
            className="link__icon"
          >
            <div
              className="icon"
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
                />
                <polyline
                  points="15 3 21 3 21 9"
                />
                <line
                  x1="10"
                  x2="21"
                  y1="14"
                  y2="3"
                />
              </svg>
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      className="about__info-item"
    >
      <p
        className="paragraph paragraph--bold"
      >
        Desenvolvido por:
      </p>
      <p
        className="paragraph"
      >
        Giovane de Oliveira
      </p>
      <p
        className="paragraph"
      >
        giovane.oliveira@softexpert.com
      </p>
    </div>
    <div
      className="about__info-item"
    >
      <p
        className="paragraph paragraph--bold"
      >
        Bugs e/ou melhorias? Cadastre uma issue em:
      </p>
      <button
        className="link"
        onClick={[Function]}
      >
        <div
          className="link__text"
        >
          BuscaTermos - Issues
        </div>
        <div
          className="link__icon"
        >
          <div
            className="icon"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
              />
              <polyline
                points="15 3 21 3 21 9"
              />
              <line
                x1="10"
                x2="21"
                y1="14"
                y2="3"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
  </div>
  <div
    className="about__button"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      <div
        className="button__content"
      >
        Voltar
      </div>
    </button>
  </div>
</div>
`;

exports[`About snapshots should render without app update 1`] = `
<div
  className="about"
>
  <div
    className="about__title"
  >
    <h1
      className="heading-1"
    >
      BuscaTermos
    </h1>
    <div
      className="about__title-date"
    >
      <p
        className="paragraph"
      >
        2014 - 2022
      </p>
    </div>
  </div>
  <div
    className="about__info"
  >
    <div
      className="about__info-item"
    >
      <p
        className="paragraph paragraph--bold"
      >
        Versão:
      </p>
      <p
        className="paragraph"
      >
        1.6.9
      </p>
    </div>
    <div
      className="about__info-item"
    >
      <p
        className="paragraph paragraph--bold"
      >
        Atualizações:
      </p>
      <div
        className="about__info-updates"
      >
        <div
          className="about__info-updates-icon"
        >
          <div
            className="icon"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="20 6 9 17 4 12"
              />
            </svg>
          </div>
        </div>
        <p
          className="paragraph"
        >
          O aplicativo está atualizado
        </p>
      </div>
    </div>
    <div
      className="about__info-item"
    >
      <p
        className="paragraph paragraph--bold"
      >
        Desenvolvido por:
      </p>
      <p
        className="paragraph"
      >
        Giovane de Oliveira
      </p>
      <p
        className="paragraph"
      >
        giovane.oliveira@softexpert.com
      </p>
    </div>
    <div
      className="about__info-item"
    >
      <p
        className="paragraph paragraph--bold"
      >
        Bugs e/ou melhorias? Cadastre uma issue em:
      </p>
      <button
        className="link"
        onClick={[Function]}
      >
        <div
          className="link__text"
        >
          BuscaTermos - Issues
        </div>
        <div
          className="link__icon"
        >
          <div
            className="icon"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
              />
              <polyline
                points="15 3 21 3 21 9"
              />
              <line
                x1="10"
                x2="21"
                y1="14"
                y2="3"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
  </div>
  <div
    className="about__button"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      <div
        className="button__content"
      >
        Voltar
      </div>
    </button>
  </div>
</div>
`;
